codegenerator/cli/templates/static/codegen/src/eventFetching/PartitionedFetchState.res

```diff
@@ -214,57 +214,38 @@ let getMostBehindPartitions = (
   ->Js.Array.slice(~start=0, ~end_=maxNumQueries)
 }
 
-let updatePartition = (self: t, ~fetchState: FetchState.t, ~partitionId: partitionIndex) => {
-  {...self, partitions: self.partitions->Utils.Array.setIndexImmutable(partitionId, fetchState)}
-}
-
 type nextQueries = WaitForNewBlock | NextQuery(array<FetchState.nextQuery>)
+
 /**
 Gets the next query from the fetchState with the lowest latestFetchedBlock number.
 */
-let getNextQueriesOrThrow = (
-  self: t,
-  ~currentBlockHeight,
-  ~maxPerChainQueueSize,
-  ~partitionsCurrentlyFetching,
-) => {
-  let optUpdatedPartition = ref(None)
-  let includesWaitForNewBlock = ref(false)
+let getNextQueries = (self: t, ~maxPerChainQueueSize, ~partitionsCurrentlyFetching) => {
+  let partitionsCopy = self.partitions->Js.Array2.copy
   let nextQueries = []
+
   self
   ->getMostBehindPartitions(
     ~maxNumQueries=Env.maxPartitionConcurrency,
     ~maxPerChainQueueSize,
     ~partitionsCurrentlyFetching,
   )
   ->Array.forEach(({fetchState, partitionId}) => {
-    switch fetchState->FetchState.getNextQuery(~currentBlockHeight, ~partitionId) {
-    | Ok((nextQuery, optUpdatesFetchState)) =>
-      switch nextQuery {
-      | NextQuery(q) => nextQueries->Js.Array2.push(q)->ignore
-      | WaitForNewBlock => includesWaitForNewBlock := true
-      | Done => ()
-      }
-      switch optUpdatesFetchState {
-      | Some(fetchState) =>
-        optUpdatedPartition :=
-          optUpdatedPartition.contents
-          ->Option.getWithDefault(self)
-          ->updatePartition(~fetchState, ~partitionId)
-          ->Some
-      | None => ()
-      }
-    | Error(e) =>
-      e->ErrorHandling.mkLogAndRaise(~msg="Unexpected error getting next query in partition")
+    let mergedFetchState = fetchState->FetchState.mergeRegistersBeforeNextQuery
+    partitionsCopy->Js.Array2.unsafe_set(partitionId, mergedFetchState)
+
+    switch mergedFetchState->FetchState.getNextQuery(~partitionId) {
+    | Done => ()
```

Comment:
In theory there might be an issue when `getMostBehindPartitions` returns us some partitions which are all Done, and in this case the queries will be empty which will stop the indexer. I don't know how real it is though.

Reply:
As a solution we probably should slice partitions exceeding the queue when we append a new query, instead of inside of the `getMostBehindPartitions` function